creusot-contracts/src/logic/fmap.rs

```diff
@@ -291,11 +291,12 @@ impl<K, V: ?Sized> FMap<K, V> {
     #[ensures(if self.contains(*key) {
             match result {
                 None => false,
-                Some(r) => *(*self).lookup_unsized(*key) == *r &&
+                Some(r) => (^self).contains(*key) &&
+                           *(*self).lookup_unsized(*key) == *r &&
                            *(^self).lookup_unsized(*key) == ^r,
```

Comment:
Shouldn't contains be implied by that?

Reply:
I don't think so... First the following function
```rust
#[requires(map.contains(*key))]
#[ensures(forall<k: K> (*map).contains(k) == (^map).contains(k))]
pub fn foo<K, V>(map: &mut FMap<K, V>, key: &K) {
    let _ = map.get_mut_ghost(key);
}
```
cannot be proved automatically with the current spec.

But beyond that, `lookup_unsized` does not make the assumption that the key is present in the map, it just specify the behavior in the case it _is_ present.

Reply:
> But beyond that, lookup_unsized does not make the assumption that the key is present in the map, it just specify the behavior in the case it is present.

Right then your spec makes sense :+1: